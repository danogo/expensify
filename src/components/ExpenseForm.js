import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';


// expense form is reused in AddExpensePage and EditExpensePage
export default class ExpenseForm extends Component {
  // can't use this.props within class property state, state has to be inside constructor to use props
  constructor(props) {
    super(props);
    // props.expense ? editing expense : creating new expense
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    }
    console.log(props)
  }

  handleDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  handleAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  handleNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  handleDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  
  handleCalendarFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  handleFormSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount for the expense.'}));
    } else {
      this.setState(() => ({ error: ''}));
      // valueOf to get timestamp in milliseconds from moment object
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf()
      })
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.handleFormSubmit}>
          <input 
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.handleDescriptionChange}
          />
          <input 
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.handleAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt} 
            onDateChange={this.handleDateChange} 
            focused={this.state.calendarFocused} 
            onFocusChange={this.handleCalendarFocusChange} 
            id="calendar"
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea 
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.handleNoteChange}
          />
          <button>Add expense</button>
        </form>
      </div>
    );
  }
}