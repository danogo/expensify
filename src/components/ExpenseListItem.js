import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({description, amount, createdAt, dispatch, id}) => (
  <div>
    <h3>{description}</h3>
    <p>Amount: {amount}</p>
    <p>Created at: {createdAt}</p>
    <button onClick={(e) => {
      dispatch(removeExpense({id}));
    }}>Remove</button>
  </div>
);

// here we connect just to get dispatch method to our WrappedComponent props
export default connect()(ExpenseListItem);