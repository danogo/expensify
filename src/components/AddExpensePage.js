import React, { Component } from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

// exporting unconnected component for testing
export class AddExpensePage extends Component {
  onSubmit = expense => {
    // props.dispatch(addExpense(expense));
    this.props.addExpense(expense);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <h1>Add expense</h1>
        <ExpenseForm 
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
}

const mapDispatchToProps = dispatch => {
  return {
    addExpense: expense => {
      dispatch(addExpense(expense));
    }
  }
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);