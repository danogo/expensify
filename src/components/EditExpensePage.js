import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends Component {
  handleSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }

  handleRemove = () => {
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
      <ExpenseForm 
        expense={this.props.expense}
        onSubmit={this.handleSubmit}
      />
      <button onClick={this.handleRemove}>Remove</button>
    </div>
    );
  }
}

// props are current available props, we are adding another prop from state
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch, props) => {
  return {
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: id => dispatch(removeExpense(id))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);