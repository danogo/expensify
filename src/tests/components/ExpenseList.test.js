import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import mockExpenses from '../fixtures/expenses';

test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={mockExpenses} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList empty message when there are no expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={[]}/>);
  expect(wrapper).toMatchSnapshot();
});