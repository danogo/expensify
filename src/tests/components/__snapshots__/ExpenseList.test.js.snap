// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList empty message when there are no expenses 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={34500}
    createdAt={0}
    description="Food"
    id={1}
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={734500}
    createdAt={-345600000}
    description="Rent"
    id={2}
    key="2"
    note="bill for one month"
  />
  <ExpenseListItem
    amount={13370}
    createdAt={345600000}
    description="Book"
    id={3}
    key="3"
    note="12 rules of life"
  />
</div>
`;
